### PR TITLE
merge valid punctual light culling for v3.4.0

### PR DESCRIPTION
**Pull request by troublemaker52025**
Re: cocos-creator/3d-tasks#

Changelog:
 *

### CODE REVIEW

**Comment by troublemaker52025** [comment]
* https://github.com/cocos-creator/3d-tasks/issues/9356

**Comment by troublemaker52025** [comment]
该优化需要同步到原生

File: cocos/core/pipeline/pipeline-scene-data.ts [hunk 1]
```diff
@@ -86,6 +86,7 @@ export class PipelineSceneData {
       * @en The list for render objects, only available after the scene culling of the current frame.
       * @zh 渲染对象数组，仅在当前帧的场景剔除完成后有效。
       */
+    public validPunctualLights: Light[] = [];
```
Comment (pandamicro) [inline cocos/core/pipeline/pipeline-scene-data.ts]: Inserted to wrong place, not matching the comment

File: cocos/core/pipeline/render-additive-light-queue.ts [hunk 2]
```diff
@@ -142,8 +140,6 @@ export class RenderAdditiveLightQueue {
         this._instancedQueue.clear();
         this._batchedQueue.clear();
 
-        this._validLights.length = 0;
```
Comment (pandamicro) [inline cocos/core/pipeline/render-additive-light-queue.ts:145]: `_validPunctualLights` should also be cleared

File: cocos/core/pipeline/render-additive-light-queue.ts [hunk 3]
```diff
@@ -171,13 +167,10 @@ export class RenderAdditiveLightQueue {
...
 
     public gatherLightPasses (camera: Camera, cmdBuff: CommandBuffer) {
-        const validLights = this._validLights;
-
         this.clear();
 
-        this._gatherValidLights(camera, validLights);
-
-        if (!validLights.length) { return; }
+        this._validPunctualLights = this._pipeline.pipelineSceneData.validPunctualLights;
```
Comment (pandamicro) [inline cocos/core/pipeline/render-additive-light-queue.ts]: If `this._validPunctualLight` always refer to pipeline scene data, there shouldn't be a private property here, we should always retrieve from `pipeline.pipelineSceneData.validPunctualLights`.

File: cocos/core/pipeline/shadow/shadow-flow.ts [hunk 4]
```diff
@@ -189,13 +204,44 @@ export class ShadowFlow extends RenderFlow {
         shadowFrameBufferMap.set(light, shadowFrameBuffer);
     }
 
+    private lightCollecting (maxReceived: number) {
+        _validLights.length = 0;
```
Comment (pandamicro) [inline cocos/core/pipeline/shadow/shadow-flow.ts]: Better to clear the array when finished using it

**Comment by troublemaker52025** [comment]
* https://github.com/cocos-creator/engine-native/pull/4048

File: cocos/core/pipeline/shadow/shadow-flow.ts [hunk 5]
```diff
@@ -90,10 +91,24 @@ export class ShadowFlow extends RenderFlow {
...
+
+            const shadowFrameBuffer = shadowFrameBufferMap.get(mainLight);
+            for (let i = 0; i < this._stages.length; i++) {
+                const shadowStage = this._stages[i] as ShadowStage;
+                shadowStage.setUsage(globalDS, mainLight, shadowFrameBuffer!);
+                shadowStage.render(camera);
+            }
+        }
+
         for (let l = 0; l < validLights.length; l++) {
```
Comment (pandamicro) [inline cocos/core/pipeline/shadow/shadow-flow.ts]: how about 

```
let validLights = this._pipeline.pipelineSceneData.validPunctualLights;
for (let l = 0; l < shadowInfo.maxReceived;) {
    const light = validLights[i];
    if (light.type === LightType.SPOT) {
        ...
        l++;
    }
}
```

`const validLights = this.lightCollecting(shadowInfo.maxReceived);` doesn't seem necessary

File: cocos/core/pipeline/render-additive-light-queue.ts (hunk 2 above)
Comment (troublemaker52025) [reply cocos/core/pipeline/render-additive-light-queue.ts:145]: 这里是对 `pipelineSceneData.validPunctualLights` 的引用。cleared 会导致引用的 `validPunctualLights` 数组被清空，没有灯光效果

File: cocos/core/pipeline/shadow/shadow-flow.ts (hunk 5 above)
Comment (troublemaker52025) [reply cocos/core/pipeline/shadow/shadow-flow.ts]: 目前不支持 sphereLight 产生的阴影，如果强行加入到数组中，会创建多余的frameBuffer造成额外的性能损耗

Comment (troublemaker52025) [reply cocos/core/pipeline/shadow/shadow-flow.ts]: 这样写会 `validLights[l]` 会越界，light.type === undefine 然后卡住

Comment (pandamicro) [reply cocos/core/pipeline/shadow/shadow-flow.ts]: 多加一个条件呀

Comment (troublemaker52025) [reply cocos/core/pipeline/shadow/shadow-flow.ts]: 如果 `if (light.type === LightType.SPOT)` 条件不成立， `const light = validLights[l];` 就没有步进，原地踏步死循环。会越改越复杂

**Review by pandamicro — APPROVED** [review]

**Review by jk20012001 — APPROVED** [review]

**Pull request merged**